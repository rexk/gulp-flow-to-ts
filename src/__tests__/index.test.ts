import * as stream from 'stream';
import * as path from 'path';
import * as gulp from 'gulp';
import * as fs from 'vinyl-fs';
import * as through2 from 'through2';
import FlowToJsonPlugin from '../index';

function toMatchExtname(extname: string) {
  return through2.obj((file, enc, callback) => {
    expect(path.extname(file.path)).toBe(extname);
    expect(file.contents.toString()).toMatchSnapshot();
    callback(null, file);
  });
}

const fixture = (str: string) =>
  path.join(__dirname, '..', '__fixtures__', str);

describe('FlowToJsonPlugin', () => {
  const tests: [string, string, string][] = [
    ['should convert to jsx', fixture('ReactComponent.js'), '.tsx'],
    ['should convert to ts', fixture('Es6Js.js'), '.ts'],
    ['should remove flow comment', fixture('BlockFlow.js'), '.tsx'],
    ['should convert nullable', fixture('Nullable.js'), '.ts'],
    ['should convert import type statement', fixture('ImportType.js'), '.tsx'],
    ['should convert React.Node to ReactNode', fixture('ReactNode.js'), '.tsx'],
    ['should convert mixed to any', fixture('MixToAny.js'), '.ts']
  ];

  tests.forEach(t => {
    const [message, src, extname] = t;
    test(message, done => {
      gulp
        .src(src)
        .pipe(FlowToJsonPlugin())
        .pipe(toMatchExtname(extname))
        .on('error', done)
        .on('finish', done);
    });
  });
});
