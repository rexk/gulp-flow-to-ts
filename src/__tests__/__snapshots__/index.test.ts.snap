// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowToJsonPlugin should convert React.Node to ReactNode 1`] = `
"import * as React from 'react';

export type MyElem = React.ReactNode;

type Props = {
  age: number;
  icon: React.Node;
};

class MyNumber extends React.Component {
  render() {
    return (
      <div>
        {this.props.age} {this.props.icon}
      </div>
    );
  }
}

export default MyNumber;
"
`;

exports[`FlowToJsonPlugin should convert import type statement 1`] = `
"import { ComponentType } from 'react';

export type MyComp = ComponentType;
"
`;

exports[`FlowToJsonPlugin should convert mixed to any 1`] = `
"function MixToAny(some: any): any {
  return some;
}

export { MixToAny };
"
`;

exports[`FlowToJsonPlugin should convert nullable 1`] = `
"/**
 * Nullable test
 */
function Nullable(something?: void | stirng): ?string {
  return something;
}

export type P = {
  name?: string;
  ok?: void | boolean;
  age?: void | number;
}
"
`;

exports[`FlowToJsonPlugin should convert to jsx 1`] = `
"import React from 'react';

class ReactComponent extends React.Component {
  render() {
    return <div>React Component</div>;
  }
}

export default ReactComponent;
"
`;

exports[`FlowToJsonPlugin should convert to ts 1`] = `
"class Es6Class {
  name: string;
  constructor(name: string) {
    this.name = name;
  }
}

export default Es6Class;
"
`;

exports[`FlowToJsonPlugin should remove flow comment 1`] = `
"import React from 'react';

type Props = {
  ok: string;
};

class BlockFlow extends React.Component<Props> {
  render() {
    return <div>{this.props.ok}</div>;
  }
}

export default BlockFlow;
"
`;
